src/main/java/io/github/bobocodebreskul/context/registry/AnnotatedBeanDefinitionReader.java

```diff
@@ -100,58 +102,84 @@ public final BeanDefinitionRegistry getBeanDefinitionRegistry() {
+  private <T> void doRegisterBean(Class<T> beanClass, String beanName) {
+    log.debug("doRegisterBean method invoked: beanClass={}, name={}", beanClass.getName(), beanName);
     log.info("Registering the bean definition of class {}", beanClass.getName());
     // todo: create beanDefinitionValidator that validate things such as:
     //  bean name validity (not allowed characters), check for circular dependency
 
     var annotatedBeanDefinition = new AnnotatedGenericBeanDefinition(beanClass);
-    name = name != null ? name : generateBeanName(annotatedBeanDefinition, beanDefinitionRegistry);
-    annotatedBeanDefinition.setName(name);
+    beanName = beanName != null ? beanName : generateBeanName(annotatedBeanDefinition, beanDefinitionRegistry);
+    annotatedBeanDefinition.setName(beanName);
 
-    if (beanDefinitionRegistry.isBeanNameInUse(name)) {
+    if (beanDefinitionRegistry.isBeanNameInUse(beanName)) {
       log.error("The specified bean name is already in use");
       throw new DuplicateBeanDefinitionException(
-        "The bean definition with specified name %s already exists".formatted(name));
+          "The bean definition with specified name %s already exists".formatted(beanName));
     }
 
-    annotatedBeanDefinition.setScope(BeanDefinition.SINGLETON_SCOPE);
 
     if (ReflectionUtils.isAnnotationPresentForClass(Primary.class, beanClass)) {
-      log.trace("Found @Primary annotation on the beanName={}", name);
+      log.trace("Found @Primary annotation on the beanName={}", beanName);
       annotatedBeanDefinition.setPrimary(true);
     }
 
-    Constructor<?> beanConstructor = findBeanInitConstructor(beanClass, name);
+    annotatedBeanDefinition.setScope(
+        getBeanDefinitionScope(beanClass, beanName, annotatedBeanDefinition));
+
+    Constructor<?> beanConstructor = findBeanInitConstructor(beanClass, beanName);
     log.debug("Constructor found for bean class [{}]: [{}]", beanClass.getName(), beanConstructor);
     annotatedBeanDefinition.setInitConstructor(beanConstructor);
     List<BeanDependency> dependencies = getConstructorBeanDependencies(beanConstructor);
     log.debug("{} dependencies found for beanClass={} with beanName={}",
-        dependencies.size(), beanClass.getName(), name);
+        dependencies.size(), beanClass.getName(), beanName);
     annotatedBeanDefinition.setDependencies(dependencies);
     annotatedBeanDefinition.setAutowireCandidate(
-      isBeanAutowireCandidate(beanClass, beanDefinitionRegistry));
+        isBeanAutowireCandidate(beanClass, beanDefinitionRegistry));
 
-    beanDefinitionRegistry.registerBeanDefinition(name, annotatedBeanDefinition);
+    beanDefinitionRegistry.registerBeanDefinition(beanName, annotatedBeanDefinition);
     log.trace("Registered bean definition: {}", annotatedBeanDefinition);
   }
+
+  private static <T> String getBeanDefinitionScope(Class<T> beanClass, String beanName,
```

Comment:
You don't need a parametrization type here.

Reply:
> You don't need a parametrization type here.

It is a bad practice to use raw Class. Besides that I will have to use type casting which is also not good